### ISSUE TITLE
Performance: `GetNextColor()` is slower now that it ignores certain plot types

### ISSUE THREAD

**Issue by 0xfded** [ScottPlot/ScottPlot#4218]
This isn't a bug per se, more of a significant change in performance.   Although the example program is unrealistic, the performance difference is significant enough it still may cause performance issues when using a small number of colors.  

That being said, it's unlikely the average user would encounter this, so it may not be worth addressing, I just thought I'd mention it.

It appears this was introduced in commit https://github.com/ScottPlot/ScottPlot/commit/3da89c3474f3cb459b04ea56782042e9893da249 as a fix to issue https://github.com/ScottPlot/ScottPlot/pull/4175

**Issue:** Accessing colors in a colormap is approximately 500 times slower than in the previous version

**ScottPlot Version:** ScottPlot.WPF 5.0.38

```cs
using System.Diagnostics;
using ScottPlot;

var colorMap = new ScottPlot.Colormaps.Algae();
var plot = new ScottPlot.Plot();

var startTime = Stopwatch.GetTimestamp();
var iterations = 0;
for(int count = 1; count < 256; count++)
{
	var xs = Generate.Consecutive(count);
	var ys = Generate.Repeating(count, count);
	var colors = colorMap.GetColors(count);

	for(int i=0; i<count; i++)
	{
		var circle = plot.Add.Circle(xs[i], ys[i], 0.45);
		circle.FillColor = colors[i];
		circle.LineWidth = 0;
		iterations++;
	}
}

var totalSeconds = Stopwatch.GetElapsedTime(startTime).TotalSeconds;
Console.WriteLine($"{iterations:N0} iterations in {totalSeconds:N2} seconds.   {iterations/totalSeconds:N2} iterations/sec");

plot.SavePng("colorMapTest.png", 1080, 1080);
```

To reproduce, run the above with version 5.0.38, then with 5.0.37.

Results: 
  
5.0.37 -  32,640 iterations in 0.01 seconds.   3,449,952.44 iterations/sec
5.0.38 - 32,640 iterations in 5.16 seconds.   6,330.13 iterations/sec

**Comment by swharden** [comment]
Hi @0xfded, thanks for the report and detailed code example!

You're right that it is a pretty edge case (it only slows down plots that have called `Add()` methods many thousands of times, but it seems worth improving.

Does changing:
```cs
List<Type> PlottablesThatDoNotGetColors
```
to
```cs
HashSet<Type> PlottablesThatDoNotGetColors
```

significantly improve performance?

If so, I'd be happy to review a pull request to merge this in!

If not (or also) an interesting alternative could be to stop doing this costly type lookup if the plot contains more than 100 items. In such a situation it would be assumed that users are adding things in a loop, and skipping colors probably wouldn't matter with that many objects anyway.

**Comment by StendProg** [comment]
HashSet increases performance by 2x. But it can never cancel the linear growth of `GetNextColor()` calculation complexity.

One of the solutions may be to allow user change `PlottableAdder` behavior. For example setting the `FixedColor` property. Which cancels the `GetNextColor()` calculation.

```c#
public class PlottableAdder
{
  public Color FixedColor? {get;set;}= null;
  ..

}

public Color GetNextColor()
{
    if (FixedColor.HasValue)
       return FixedColor.Value;
  
    int coloredPlottableCount = Plot.PlottableList.Where(x => !PlottablesThatDoNotGetColors.Contains(x.GetType())).Count();
    return Palette.GetColor(coloredPlottableCount);

}
```

Or we could try introducing a special one-color palette so we don't have to introduce a new property.

```c#
public Color GetNextColor()
{
    if (Pallete is FixedColorPallete)
       return Pallete.GetColor(0);
  
    int coloredPlottableCount = Plot.PlottableList.Where(x => !PlottablesThatDoNotGetColors.Contains(x.GetType())).Count();
    return Palette.GetColor(coloredPlottableCount);

}
```
```

**Comment by swharden** [comment]
> HashSet increases performance by 2x. But it can never cancel the linear growth of GetNextColor() calculation complexity

This is a very good point!

The solutions proposed are interesting, but fundamentally I think the problem is that `GetNextColor()` is using the number of plottables and perhaps that isn't the best way.

I'm curious if a simpler strategy could work -- increment a counter by one every time `GetNextColor()` is called? Perhaps the counter could be reset to zero if the plot is cleared (or the number of plottables is zero). I'll give this a try!

**Comment by swharden** [comment]
I implemented that strategy in #4230 and it works surprisingly well! The plot types that we don't want getting counted don't call GetNextColor(), so they don't increment the counter:  PolarAxis, Annotation, Benchmark, CandlestickPlot, OhlcPlot, Heatmap, ImageMarker, ImageRect, IsoLines, Pie, Text
